repro/doc/repro.8
.TH repro 8 "February 2006"
.\" ====================================================================
.\" The Vovida Software License, Version 1.0 
.\" 
.\" Copyright (c) 2000 Vovida Networks, Inc.  All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in
.\"    the documentation and/or other materials provided with the
.\"    distribution.
.\" 
.\" 3. The names "VOCAL", "Vovida Open Communication Application Library",
.\"    and "Vovida Open Communication Application Library (VOCAL)" must
.\"    not be used to endorse or promote products derived from this
.\"    software without prior written permission. For written
.\"    permission, please contact vocal@vovida.org.
.\"
.\" 4. Products derived from this software may not be called "VOCAL", nor
.\"    may "VOCAL" appear in their name, without prior written
.\"    permission of Vovida Networks, Inc.
.\" 
.\" THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESSED OR IMPLIED
.\" WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, TITLE AND
.\" NON-INFRINGEMENT ARE DISCLAIMED.  IN NO EVENT SHALL VOVIDA
.\" NETWORKS, INC. OR ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT DAMAGES
.\" IN EXCESS OF $1,000, NOR FOR ANY INDIRECT, INCIDENTAL, SPECIAL,
.\" EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
.\" PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
.\" PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
.\" OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
.\" USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
.\" DAMAGE.
.\" 
.\" ====================================================================
.\" 
.\" This software consists of voluntary contributions made by Vovida
.\" Networks, Inc. and many individuals on behalf of Vovida Networks,
.\" Inc.  For more information on Vovida Networks, Inc., please see
.\" <http://www.vovida.org/>.
.\"
.\"
.SH NAME
repro \- repro SIP proxy server
.SH SYNOPSIS
.B
repro [OPTIONS...]

.SH DESCRIPTION
.B repro
is an open-source, free SIP server. SIP is changing the way people communicate
using the Internet. It is not only about making phone calls over the Net. The
SIP protocol and its extensions defines the way of establishing, modifying and
ending interactive sessions, no matter if they are voice, video, IM or a
combination of them. At the heart of SIP architecture, there are certain
services which needs to be provided at some place of the network.
.B repro
provides SIP proxy, registrar, redirect, and identity services. These services
are the foundation needed to run a SIP service.

.SH OPTIONS
.TP
\fB\-l\fR, \fB\-\-log\-type\fR=\fBcout\fR|\fBsyslog\fR|\fBcerr\fR
Set where to send logging messages. The default \fBcout\fR sends the
messages to standard output, \fBcerr\fR uses the standard error pipe.
With \fBsyslog\fR the messages are send to the system log using the logging
facility \fBlocal6\fR. See
.BR syslogd (8)
or
.BR syslog.conf (5)
for more information.
.TP
\fB\-v\fR, \fB\-\-log\-level\fR=\fBSTACK\fR|\fBDEBUG\fR|\fBINFO\fR|\fBWARNING\fR|\fBALERT\fR
Set the minimum level a logging message needs to have in order to by
written. The severity rises from left to right in the above order. The
default is to log messages with level \fBINFO\fR or above.
.TP
\fB\-r\fR, \fB\-\-record\-route\fR=\fIsip:example.com\fR
Activate record routing. The argument is the URI used in the Record-Route
header field.
.TP
\fB\-\-udp\fR=\fI5060\fR
Specify the port to use for the UDP transport. Set to 0 to turn
UDP off. Defaults to port 5060.
.TP
\fB\-\-tcp\fR=\fI5060\fR
Specify the port to use for the TCP transport. Set to 0 to turn TCP
off. Defaults to port 5060.
.TP
\fB\-t\fR, \fB\-\-tls\-domain\fR=\fIexample.com\fR
Act as a TLS server for specified domain.
.TP
\fB\-\-tls\fR=\fI5061\fR
Set the port to use for the TLS transport. Set to 0 to turn TLS off.
Defaults to port 5061.
.TP
\fB\-\-dtls\fR=\fI0\fR
Set the port to use for the DTLS transport. Set to 0 to turn DTLS off
which is the default.
.TP
\fB\-\-enable\-cert\-server
Run a cert server.
.TP
\fB\-c\fR, \fB\-\-cert\-path\fR=\fISTRING\fR
Set the path for certificates. Defaults to \fB~/.sipCerts\fR.
.TP
\fB\-\-enable\-v6\fR
Enable IPv6.
.TP
\fB\-\-disable\-v4\fR
Disable IPv4.
.TP
\fB\-\-disable\-auth\fR
Disable DIGEST challenges for certain SIP requests.
.TP
\fB\-\-disable\-web\-auth\fR
Disable authentication for the web administration server.
.TP
\fB\-\-disable\-reg\fR
Disable the registrar.
.TP
\fB\-i\fR, \fB\-\-interfaces\fR=\fIsip:10.1.1.1:5065;transport=tls\fR
Specify interfaces to add transports to. Each transport is given as a
SIP-URI with the IP address and port of the local interface as the
hostport part and the transport protocol as a \fBtransport\fR parameter.
Several transports are separated by a comma.

For example, to set up two transports, one for TLS and one for UDP use
.B sip:192.168.1.200:5060;transport=tls,sip:192.168.1.200:5060;transport=udp
.TP
\fB\-d\fR, \fB\-\-domains\fR=\fIexample.com,foo.com\fR
Specify the list of domains this proxy is authorative for separated by
comma.
.TP
\fB\-R\fR, \fB\-\-route\fR=\fIsip:p1.example.com,sip:p2.example.com\fR
Specify a route set where all requests this proxy is authorative for are
sent to. Using this option overides the routes set through the web
interface.
.TP
\fB\-\-reqChainName\fR=\fISTRING\fR
Name of request chain used for processing requests. Currently, the only
chain available is \fBdefault\fR.
.TP
\fB\-\-http\fR=\fI5080\fR
Specify the port used by the HTTP server. Defaults to 5080.
.TP
\fB\-\-recursive\-redirect\fR
Enable to handle of 3xx responses in the proxy.
.TP
\fB\-\-q\-value\fR
Enable q-value processing. The q-value can be given in registrations and is
used by the location server to order forwarding for multiple registrations.
Without this option a request is forked to all registrations.
.TP
\fB\-\-q\-value\-behavior\fR=\fBFULL_SEQUENTIAL\fR|\fBEQUAL_Q_PARALLEL\fR|\fBFULL_PARALLEL\fR
Specify forking behavior if \fB\-\-q\-value\fR is given.
With \fBFULL_SEQUENTIAL\fR one target is called after another in the order
of their q values. Using \fBEQUAL_Q_PARALLEL\fR (the default), the request
is sent in
batches of all registrations with the same q value. The batches are again
ordered by the q value.
\fBFULL_PARALLEL\fR causes the request to be sent in parallel to all
registrations.
.TP
\fB\-\-q\-value\-cancel\-btw\-fork\-groups\fR
If given, groups of parallel forks are canceled after the period specified
by the \fB\-\-q\-value\-ms\-before\-cancel\fR option if no response was
received.
.TP
\fB\-\-q\-value\-wait\-for\-terminate\-btw\-fork\-groups\fR
If given, the proxy waits for groups of parallel forks to terminate before
a new group is started. Otherwise, the next group is started once the
waiting time has passed.
.TP
\fB\-\-q\-value\-ms\-between\-fork\-groups\fR=\fIINT\fR
Specify the number of milliseconds to wait for a response from a group of
parallel forks before starting another group.
.TP
\fB\-\-q\-value\-ms\-before\-cancel\fR=\fIINT\fR
Specify the number of milliseconds to wait before cancelling a group of
parallel forks if the \fB\-\-q\-value\-cancel\-btw\-fork\-groups\fR option
is given.
.TP
\fB\-e\fR, \fB\-\-enum\-suffix\fR=\fIe164.arpa\fR
Specify enum suffix to search.
.TP
\fB\-b\fR, \fB\-\-allow\-bad\-reg\fR
Allow To tag in registrations.
.TP
\fB\-\-timer\-C\fR=\fI180\fR
Specify the length of timer C in sec which specifies the time a proxy
waits before generating a timeout response. Set to 0 or a negative value
to turn timer C of completely.
.TP
\fB\-\-admin\-password\fR=\fISTRING\fR
Set the web administrator password.
.PP
Help options:
.TP
\fB\-?\fR, \fB\-\-help\fR
Show a help message.
.TP
\fB\-\-usage
Display brief usage message.
.TP
\fB\-\-version\fR
Display the version information.

.SH SEE ALSO
Repro web site at
.B http://www.resiprocate.org/About_Repro

.\".SH AUTHORS

.\".SH BUGS





repro/repro.cxx

#if defined(HAVE_CONFIG_H)
  #include "config.h"
#endif

#include <signal.h>
#include "repro/ReproRunner.hxx"
#include "rutil/Socket.hxx"
#include "rutil/Log.hxx"
#include "rutil/Logger.hxx"

#include "rutil/WinLeakCheck.hxx"

#define RESIPROCATE_SUBSYSTEM resip::Subsystem::REPRO

using namespace repro;
using namespace resip;
using namespace std;

static bool finished = false;
static bool receivedHUP = false;

static void
signalHandler(int signo)
{
#ifndef _WIN32
   if(signo == SIGHUP)
   {
      InfoLog(<<"Received HUP signal, logger reset");
      Log::reset();
      receivedHUP = true;
      return;
   }
#endif
   std::cerr << "Shutting down" << endl;
   finished = true;
}

/*
   Extending Repro by adding custom processors to the chain is as easy as overriding one of the 
   ReproRunner class virtual methods:
   virtual void addProcessor(repro::ProcessorChain& chain, std::auto_ptr<repro::Processor> processor);
   virtual void makeRequestProcessorChain(repro::ProcessorChain& chain);
   virtual void makeResponseProcessorChain(repro::ProcessorChain& chain);
   virtual void makeTargetProcessorChain(repro::ProcessorChain& chain);

   Override the makeXXXProcessorChain methods to add processors to the beginning or end of any chain,
   or override the addProcessor method, and you can examine the name of the processor being 
   added and add your own process either before or after the correct processor.  

   WARNING: Be careful when checking for names of optional processors.  Depending on
            the configuration some processors may not be enabled.

   Create an instance of your overridden ReproRunner class and call run to start everything 
   up.

   Example:

   class MyCustomProcessor : public Processor
   {
      public:
         MyCustomProcessor(ProxyConfig& config) : Processor("MyCustomProcessor") {}
         virtual ~MyCustomProcessor() {}
      
         virtual processor_action_t process(RequestContext &context)
         {  
            DebugLog(<< "Monkey handling request: " << *this << "; reqcontext = " << context);
            
            ...Do something interesting here....

            return Processor::Continue;
         }
      }
   };

   class MyReproRunner : public ReproRunner
   {
   public:
      MyReproRunner() {}
      virtual ~MyReproRunner() {}
   
   protected:
      virtual void addProcessor(repro::ProcessorChain& chain, std::auto_ptr<repro::Processor> processor)
      {
         if(processor->getName() == "LocationServer")
         {
            // Add MyCustomProcessor before LocationServer
            addProcessor(chain, std::auto_ptr<Processor>(new MyCustomProcessor(*mProxyConfig)));
         }
         ReproRunner::addProcessor(chain, processor);  // call base class implementation
      }
   };

   There is also generic storage available for custom Processor at three different
   scope levels via a KeyValueStore:
     -Global Proxy Scope - Proxy::getKeyValueStore
     -Request Scope - RequestContext::getKeyValueStore
     -Target Scope - Target::getKeyValueStore
   Before this storage can be used you must statically allocate a storage key.
   See mFromTrustedNodeKey use in the IsTrustedNode class for an example.

*/

int
main(int argc, char** argv)
{
   // Install signal handlers
#ifndef _WIN32
   if ( signal( SIGPIPE, SIG_IGN) == SIG_ERR)
   {
      cerr << "Couldn't install signal handler for SIGPIPE" << endl;
      exit(-1);
   }
   if ( signal( SIGHUP, signalHandler ) == SIG_ERR )
   {
      cerr << "Couldn't install signal handler for SIGHUP" << endl;
      exit( -1 );
   }
#endif

   if ( signal( SIGINT, signalHandler ) == SIG_ERR )
   {
      cerr << "Couldn't install signal handler for SIGINT" << endl;
      exit( -1 );
   }

   if ( signal( SIGTERM, signalHandler ) == SIG_ERR )
   {
      cerr << "Couldn't install signal handler for SIGTERM" << endl;
      exit( -1 );
   }

   // Initialize network
   initNetwork();

#if defined(WIN32) && defined(_DEBUG) && defined(LEAK_CHECK) 
   { FindMemoryLeaks fml;
#endif

   ReproRunner repro;
   if(!repro.run(argc, argv))
   {
      cerr << "Failed to start repro, exiting..." << endl;
      exit(-1);
   }

   // Main program thread, just waits here for a signal to shutdown
   while (!finished)
   {
      sleepMs(1000);
      if(receivedHUP)
      {
         repro.onHUP();
         receivedHUP = false;
      }
   }

   repro.shutdown();

#if defined(WIN32) && defined(_DEBUG) && defined(LEAK_CHECK) 
   }
#endif
   return 0;
}


/* ====================================================================
 * The Vovida Software License, Version 1.0 
 * 
 * Copyright (c) 2000-2005
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 * 
 * 3. The names "VOCAL", "Vovida Open Communication Application Library",
 *    and "Vovida Open Communication Application Library (VOCAL)" must
 *    not be used to endorse or promote products derived from this
 *    software without prior written permission. For written
 *    permission, please contact vocal@vovida.org.
 *
 * 4. Products derived from this software may not be called "VOCAL", nor
 *    may "VOCAL" appear in their name, without prior written
 *    permission of Vovida Networks, Inc.
 * 
 * THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESSED OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, TITLE AND
 * NON-INFRINGEMENT ARE DISCLAIMED.  IN NO EVENT SHALL VOVIDA
 * NETWORKS, INC. OR ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT DAMAGES
 * IN EXCESS OF $1,000, NOR FOR ANY INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
 * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
 * DAMAGE.
 * 
 * ====================================================================
 * 
 * This software consists of voluntary contributions made by Vovida
 * Networks, Inc. and many individuals on behalf of Vovida Networks,
 * Inc.  For more information on Vovida Networks, Inc., please see
 * <http://www.vovida.org/>.
 *
 */
/*
 * vi: set shiftwidth=3 expandtab:
 */
